Add more motivation to the intro (#60)
melange-re/melange-for-react-devs: docs/intro/index.md

This is a work in progress.

We'd love to hear your feedback! If you find any errors or have suggestions for
the book, please [file an
issue](https://github.com/ahrefs/melange-for-react-devs/issues) or ping @feihong
on the [#melange channel in the Reason Discord](https://discord.gg/reasonml).

:::

## What is Melange?

Melange is a collection of tools that bring the robustness of a mature,
statically-typed multi-paradigm programming language to the JavaScript
ecosystem. That language is OCaml, a language invented in the 1990s which has
been battle-tested by industry stalwarts. The heart of Melange is (1) a compiler
that translates OCaml to human-readable JavaScript and (2) built-in language
constructs for zero-cost interoperation with JavaScript.

## Why OCaml?

OCaml codebases scale well both in terms of quantity of lines and number of
contributors. The sound type system helps to prevent ambiguous behavior in your
program---if it compiles, it runs without runtime errors. OCaml and React (via

> [ReasonReact][reasonreact]) are an effective, FP-friendly combination for

Reviewer: A quick note on how ReasonReact relates to OCaml/ReasonML - it's not clear to me. May not be important at this point, but just a thought.

building frontend UIs using JSX syntax. If the backend is also written in OCaml,
you can share types between the frontend and backend, ensuring that they stay in
sync[^1].  It's even possible to write [universal React
components](https://github.com/ml-in-barcelona/server-reason-react) that are
rendered on the server with the performance of native code[^2].

All that said, how does OCaml compare against other prominent compile-to-JS
languages?

**TypeScript** is a gradually-typed superset of JavaScript. It's easier to learn
and adopt because its design highly emphasizes compatibility with JavaScript.
This brings a few footguns, such as `any`, `as` casting, `ts-nocheck` etc,
making it less type-safe than OCaml (and all the other languages on this list)[^3].

> - Zero-cost JS interop? Yes

Reviewer: Maybe for each of these lists, a quick "Pros and Cons"

programming.

- Zero-cost JS interop? No. Elm treats JavaScript as unsafe, interacting with JS
  via ports (more like message passing than traditional FFI). Ergonomic
  integration with React and other libraries isn't possible.
- Can use with React? No. Elm has its own framework with some similar concepts.
- Supports JSX? No
- Pattern matching? Yes
- Can use on server? No

**F#** is a fullstack language similar to OCaml and is strongly aligned with the
.NET ecosystem.

- Zero-cost JS interop? Probably
- Can use with React? Yes. [Feliz](https://github.com/Zaid-Ajaj/Feliz) is an F#
  wrapper library for React.
- Supports JSX? No. Feliz uses ordinary function call syntax for render logic.
- Pattern matching? Yes
- Can use on server? Yes, via .NET or Node.

**Gleam** is a fullstack language similar to OCaml which is aligned with the
BEAM ecosystem.

- Zero-cost JS interop? No, but Gleam's JS compilation target is relatively new,
  so it may become more optimized in the future.
- Can use with React? Yes, via [redraw](https://github.com/ghivert/redraw) and [react-gleam](https://github.com/brettkolodny/react-gleam).
- Supports JSX? No

> - Pattern matching? Yes

Reviewer 2: I don't see the point of this section, this is a book of Reason and Melange. 

Let's encourage people to read about comparisons in the melange documentation or a blog post.

Reviewer: I'd kind of lean this way too. It reads a bit like a sales pitch. Maybe at a minimum, cut down on a few of the alternatives.

  battle-tested as ReasonReact.
- Supports JSX? No
- Pattern matching? Yes
- Can use on server? Yes, via native compilation or Node

As you can see, each compile-to-JS language has its own design goals, which
dictate its strengths and weaknesses relative to OCaml and Melange.

If you spot any inaccuracies in this language comparison, please [file an
issue](https://github.com/melange-re/melange-for-react-devs/issues).

## What's in this book

This is a project-based, guided introduction to Melange and its ecosystem,
including ReasonReact, the dune build system, and various useful libraries.
Because Melange uses both OCaml and JavaScript ecosystems, there are quite a few
tools and concepts to learn. Therefore we try to make each chapter small and
digestible, not introducing too many things at once. At the end of each chapter,
you'll find exercises with solutions.

## Audience

You should already know how to make frontend applications in JavaScript, in
particular with [React](https://react.dev/). You should be interested in
learning how to leverage your existing knowledge to build apps using
[ReasonReact][reasonreact]. You do not need to know OCaml[^6]---we'll slowly

> introduce the basics of the language throughout the tutorial. That said, a good

Reviewer 2: probably the audience is a bit "informed" about FP

complement to this guide is [OCaml Programming: Correct + Efficient +
Beautiful](https://cs3110.github.io/textbook/), which teaches the language from

> the ground up and goes much deeper into its features.

Reviewer: I really like the external resources for further reading!

This is a work in progress.

We'd love to hear your feedback! If you find any errors or have suggestions for
the book, please [file an
issue](https://github.com/ahrefs/melange-for-react-devs/issues) or ping @feihong
on the [#melange channel in the Reason Discord](https://discord.gg/reasonml).

:::

## What is Melange?

Melange is a collection of tools that bring the robustness of a mature,
statically-typed multi-paradigm programming language to the JavaScript
ecosystem. That language is OCaml, a language invented in the 1990s which has
been battle-tested by industry stalwarts. The heart of Melange is a compiler
that translates OCaml to human-readable JavaScript and built-in language
constructs for zero-cost interoperation with JavaScript.

> ## Why OCaml?

Reviewer 2: Same discussion as ever, but "Why OCaml?" when you will write the book in reason is strange, confusing

Author: It seems to me that most of the benefits of Melange come from OCaml, and the book does teach OCaml (using Reason syntax).

Author: Oh wait, do you mean Melange in the sense of the subset of OCaml language supported by Melange?

Reviewer: To add my 2 cents: It might be beneficial to just add a quick note that communicates that:

Ocaml = general purpose programming language developed in the 90s (and from my understanding, derived from ML
ReasonML = a set of Ocaml extensions/different way of writing OCaml (this is my understanding as someone with lots of functional JavaScript experience, but not a whole lot of ML experience)

Also note that while we are talking about Ocaml here, the reader can expect to see a lot more ReasonML later.

OCaml codebases scale well both in terms of quantity of lines and number of
contributors. The sound type system helps to prevent ambiguous behavior in your

> program---if it compiles, it runs. OCaml and React (via

Reviewer 3: The "if it compiles, it runs" is a bit confusing because TypeScript also "runs if it compiles" 😄 What about 
```suggestion
program---if it compiles, it runs without runtime errors. OCaml and React (via
```
?

Author: Fixed in https://github.com/melange-re/melange-for-react-devs/pull/60/commits/8ca7939e01ed478f8754742525bf3f3072216704

[ReasonReact][reasonreact]) are an effective, FP-friendly combination for
building frontend UIs using JSX syntax. If the backend is also written in OCaml,
you can share types between the frontend and backend, ensuring that they stay in
sync[^1]. In the near future, it will be possible to write [universal React
components](https://github.com/ml-in-barcelona/server-reason-react) that are

> performantly rendered on the server using OCaml.

Reviewer 2: I wouldn't say "I the near future", it's already possible. It's just not polished. I would explain within the previous sentence: you can write a full-stack application (sharing types, components rendered on the server, etc etc)

Author: How about:

> If the backend is also written in OCaml,
you can share types between the frontend and backend, ensuring that they stay in
sync[^1]. It's even possible to write [universal React
components](https://github.com/ml-in-barcelona/server-reason-react) that are
rendered on the server with the performance of native code[^2].

The [^2] footnote will explain that it's not yet polished.

Reviewer 2: That's a better phrasing indeed

Author: Rephrased in https://github.com/melange-re/melange-for-react-devs/pull/60/commits/8ca7939e01ed478f8754742525bf3f3072216704

All that said, how does OCaml compare against other prominent compile-to-JS
languages?

**TypeScript** is a gradually-typed superset of JavaScript. It is easier to

> learn and adopt, but is less type-safe than OCaml and all other langues on this

Reviewer 3: ```suggestion
learn and adopt, but is less type-safe than OCaml and all other languages on this
```

Author: Fixed in https://github.com/melange-re/melange-for-react-devs/pull/60

> learn and adopt, but is less type-safe than OCaml and all other langues on this

Reviewer 2: I would try to be more expressive on comparing with TypeScript.

**TypeScript** is a gradually-typed superset of JavaScript. It's easier to
learn and adopt. His design, following closely JavaScript, brings a few footguns, like any, "as" casting, and ts-nocheck (to say a few) becoming less type-safe than OCaml.

Author: Added in https://github.com/melange-re/melange-for-react-devs/pull/60

list. TypeScript doesn't support pattern matching at the language level.

> **ReScript** is very similar to OCaml, even at the syntax level. Its
> RescriptReact library is roughly equivalent to our ReasonReact library. The main

Reviewer 3: ```suggestion
**ReScript** is very similar to OCaml, even at the syntax level. Its
rescript-react bindings are roughly equivalent to Melange reason-react bindings. The main
```

Author: Changed in https://github.com/melange-re/melange-for-react-devs/pull/60

differences are that ReScript's only compilation target is JS, and it lacks the
metaprogramming features of OCaml, so you lose the ability to, say, [embed CSS

> styles directly into your code](https://styled-ppx.vercel.app/).

Reviewer 3: Styled-ppx [supports ReScript](https://styled-ppx.vercel.app/getting-started/rescript), so I would rather mention the ability to write universal components as the main benefit or Melange + ReasonReact.

Author: Deleted metaprogramming mention in https://github.com/melange-re/melange-for-react-devs/pull/60

> **Elm** is similar to OCaml but leans much more strongly towards pure functional

Reviewer 3: ```suggestion
**Elm** is similar to OCaml. It leans strongly towards pure functional
```

Author: Changed in https://github.com/melange-re/melange-for-react-devs/pull/60

> programming. It cannot use React but comes with its own framework. It has many
> fewer options for JS interop. Even though Elm compiles to JS, it can't be used
> for backend development, even on Node.

Reviewer 3: ```suggestion
programming. It has fewer options for JS interop and cannot use React but comes with
its own framework. Elm can't be used for backend development, even on Node.
```

Reviewer 2: "It cannot use React but comes with its own framework".

It's more like: "Elm treats JavaScript as unsafe, with possibility to interact with it via ports. Integrations like React or Melange FFI aren't possible"

Author: Changed in https://github.com/melange-re/melange-for-react-devs/pull/60

Reviewer 4: Use can use Elm on the backend: [lamdera](https://dashboard.lamdera.app/) & [elm-pages](https://elm-pages.com/), but I don't think you can use node apis.

Reviewer 2: never understood how lamdera or elm-pages run in the backend. I asked a few times here and there and didn't get a decent answer. Just offtopic, but always curious

Author: @psb I perused lamdera and elm-page sites but couldn't get a clear idea of how they work at all, so I'm hesitant to include it in the comparison

**F#** is a fullstack language similar to OCaml and is strongly aligned with the
.NET ecosystem. F# has good support for functional programming and zero-cost
interop with C#. Feliz is an F# wrapper library for React that doesn't support
JSX.

> 

Reviewer 4: I think you can use jsx in Fable/Feliz: https://fable.io/blog/2022/2022-10-12-react-jsx.html , it's just that most people use the elm like syntax.

Author: I took a look at JSX in Fable/Feliz, but the usage doesn't seem ergonomic to me (implemented using string templates). Probably just my opinion, but I don't think it qualifies as "JSX".

**Kotlin** is a type-safe fullstack language strongly associated with the
Compose Multiplatform UI framework. Kotlin-react is a Kotlin wrapper library for
React which doesn't support JSX. Kotlin has some features in common with OCaml,
but doesn't have pattern matching.

**Dart** is a type-safe fullstack language strongly associated with the Flutter
UI framework, which supports compiling UI apps to mobile, desktop, and browser
environments. There's no Dart wrapper library for modern, functional React
components.

> **OCaml** has another JS transpiler called js_of_ocaml. Unlike Melange,

Reviewer 3: There's the Melange for X developers page that has [a jsoo section](https://melange.re/v4.0.0/melange-for-x-developers#for-js-of-ocaml-developers), not sure if worth mentioning.

Also includes subsections about ReScript and TypeScript.

Author: Add links to ReScript and jsoo detailed comparisons in https://github.com/melange-re/melange-for-react-devs/pull/60/commits/8ca7939e01ed478f8754742525bf3f3072216704

Add link to detailed TypeScript comparison in https://github.com/melange-re/melange-for-react-devs/pull/60/commits/e54eb1fc68b0547ac217c14419e0481687d789b4

> js_of_ocaml prioritizes OCaml compatibility over JavaScript interop.

Reviewer 3: I'd also mention that the story to use React is much more stable and tested in Melange than in jsoo.

Author: Add this detail in https://github.com/melange-re/melange-for-react-devs/pull/60/commits/d36abdf9c62d8637370b813b775788bae694bf53

> 

Author: Originally I had one paragraph for Rust here, but not sure if it's really in the same space as it compiles to WebAssembly and not normal JS. Should I bring it back?

Reviewer 4: Is a comparison to Gleam worthwhile?

Reviewer 4: And PureScript since it was mentioned in the issue?

Author: @psb Add comparison to Gleam in https://github.com/melange-re/melange-for-react-devs/pull/60/commits/4d8e896f0e754163d0bc62f09d94f180db4d5560

I haven't heard much about PureScript in recent years, so I don't think it meets the criteria of "prominent compile-to-JS language".

As you can see, each compile-to-JS language has its own design goals, which

> dictate its strengths and weaknesses relative to OCaml and Melange.

Reviewer 2: I would move this into a separate section or even in the melange documentation

Author: The language comparison section was added in response to https://github.com/melange-re/melange-for-react-devs/issues/58 and also another reader who responded to me via DM. But I agree it makes sense to add it to Melange docs as well.
